### PR TITLE
New: Allow parallel linting in child processes (fixes #3565)

### LINKED ISSUE #3565: Lint multiple files in parallel [$500]
This is a discussion issue for adding ability to run eslint in parallel for multiple files.

The idea is that ESLint is mostly CPU bound, not IO bound, so creating multiple threads (for machine with multiple cores) might (and probably will) increase performance in a meaningful way. The downside is that currently ESLint's codebase is synchronous. So this would require rewriting everything up to and including eslint.js to be asynchronous, which would be a major effort.

I played with this a little while ago and found a few libraries for Node that handle thread pool, including detection of number of cores available on the machine.
- [Node-threads-a-gogo](https://github.com/xk/node-threads-a-gogo) - seems pretty good, but looks dead.
- [nPool](https://github.com/inh3/nPool) - seems actively in development, but has native components (C++)
- [Node WebWorkers](https://github.com/audreyt/node-webworker-threads) - seems pretty dead too.
- [Parallel](https://github.com/adambom/parallel.js) - seems dead, and no pool implementation.
- [Node Clusters](https://nodejs.org/api/cluster.html) - not stable yet, and probably isn't going to be available on Node v0.10
- WebWorkers - seems that they are only implemented in io.js
  And there are a ton of other libraries out there for this.

If anyone had any experience writing multithreaded applications for node.js and would like to suggest alternatives or comment on the above list, please feel free.

P.S. https://www.airpair.com/javascript/posts/which-async-javascript-libraries-should-i-use
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/26284182-lint-multiple-files-in-parallel?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Gaafar**
WIP: implement parallel linting

### CODE REVIEW

**Comment by ilyavolodin** [comment]
Very nice PoC! Thanks for doing this! There is one big problem with this approach however. Functions in CLIEngine are part of the public API and we can't just change them without creating a major breaking change (which we always try to avoid, even with major releases). We would have to leave existing functions as is, and create another async version of them. We also wanted to put parallel linting behind a flag, because on some CI VPCs with a single processor, it might actually slow things down quite a bit.

**Comment by Gaafar** [comment]
@ilyavolodin Sure! I just wanted to see how if it would work. We need to find a way to introduce an async API without breaking the contract, and with minimal code duplication.

Let me know if you think of any other points to consider. Also I'm not quite sure how to test this functionality.

**Comment by ilyavolodin** [comment]
Yes, preserving API is very important. We also want to put parallel linting behind CLI flag for now (it might change in the future, but initial implementation should leave single-threaded linting as a default for now. I would suggest `-p`/`--parallel` flag). One other thing to watch out for is order of reports in formatter. That needs to stay as it currently is (as in, file order in the output).
In terms of testing, I'm not sure how to test it properly. We need tests to verify that async linting is still completing correctly. We also need some tests to verify that there are multiple processes doing the linting.

**Comment by Gaafar** [comment]
Regarding the flag, I thought that maybe they can be 2 flags: 
1. `-p`: where the value is the number of parallel processes to run. If the flag is provided without a value, a default number will be used
2. `--parallel-files`: the number of files that are linted concurrently, per process. This will be the concurrency limit passed to `bluebird.map`

**Comment by Gaafar** [comment]
one other thing that I don't know how to tackle is passing `ConfigArray`. The hack I came up with is to pass it to the child process as a serialized array, but then it loses the properties stored in `internalSlotsMap`, and then I needed to add the method `deserialize` to rebuild the plugin definitions. I don't know if there is a better way to do it, or whether it would break something else.

**Comment by Gaafar** [comment]
@ilyavolodin I cleaned up the implementation a lot and found a way to implement parallel without breaking the current sync API. Check it now

**Comment by mysticatea** [comment]
Hi. Thank you for working on this.

I have a concern -- I don't think that serializing `ConfigArray` is a good idea. Config arrays have a ton of stuff we cannot serialize. It has:

- Plugin implementations
- Parser implementations
- `Error` objects about loading errors; stack trace is important to debug.
- Specific class instances such as `Minimatch`.
- etc...

Instead, I think that to load configs in workers is better. The config factory has the cache, so it will be fast than serializing the config of every file individually. It's a common case that some hundreds of files use the same config.

**Comment by Gaafar** [comment]
Actually I'm not happy either with serializing the config but I don't know how to load the config in the workers. Can you suggest how to do that?

**Comment by mysticatea** [comment]
Yes, use [CascadingConfigArrayFactory](https://github.com/eslint/eslint/blob/ba8c2aa0154561fbeca33db0343cb39a7fbd9b4f/lib/cli-engine/cascading-config-array-factory.js).

**Comment by mysticatea** [comment]
The instantiation example of the factory is here: [lib/cli-engine/cli-engine.js#L552-L561](https://github.com/eslint/eslint/blob/ba8c2aa0154561fbeca33db0343cb39a7fbd9b4f/lib/cli-engine/cli-engine.js#L552-L561). The constructor arguments are the options of `CLIEngine`. Then call `factory.getConfigArrayForFile(filePath)`.

Ah.
I have noticed that we cannot support parallel linting if a user used `cliEngine.addPlugin()`. We cannot pass the plugin implementation to workers. This should be written in the documentation as a known limitation.

**Comment by Gaafar** [comment]
Done. Much cleaner now. Thanks for the feedback! https://github.com/eslint/eslint/pull/12191/files#diff-c50f7c7af8213ec603fb51a7ee2b3cf8R907

File: lib/cli-engine/cli-engine.js [hunk 1]
```diff
@@ -804,24 +875,110 @@ class CLIEngine {
...
+            configArrayFactory,
+            options: {
+                allowInlineConfig,
+                cwd,
+                fix,
+                reportUnusedDisableDirectives
+            }
+        } = internalSlotsMap.get(this);
+
+        const verifyFilesResults = await bluebird.map(chunkFiles, async({ filePath, ignored }) => {
```
Comment (kaicataldo) [inline lib/cli-engine/cli-engine.js]: Is there a way we can do this using native Promises instead of introducing a new promise library?

Comment (Gaafar) [reply lib/cli-engine/cli-engine.js]: using bluebird mainly for the concurrency limit, to avoid trying to require all files immediately

**Comment by Gaafar** [comment]
Added an example for tests, the idea is to do each assertion once for sync API and once for async

**Comment by Gaafar** [comment]
Can someone recommend a project with thousands of files to use for benchmarking?

**Comment by ilyavolodin** [comment]
I'm not sure which project would be considered as largest, but Node core should be large enough for a benchmark, I think.

**Comment by Gaafar** [comment]
@ilyavolodin @mysticatea any ideas on the testing approach before I add other tests?

**Comment by ilyavolodin** [comment]
@Gaafar I'm not really sure how to approach unit testing in this case. I haven't done any unit tests for multi-process setup before. I know we should verify that output is the same in both cases (i.e. reports are ordered the same way in parallel linting and single-threaded).

**Comment by Gaafar** [comment]
Yeah, that's what I'm doing now, running the same assertion on the sync result and the parallel result

**Comment by ilyavolodin** [comment]
@Gaafar Could you post here some comparisons in terms of the performance with and without parallel linting?

**Comment by Gaafar** [comment]
@ilyavolodin I'm waiting to finalize a few things then post the comparison. So far in my tests I can see parallel is around 20% faster on 2 cpu cores and 40% on 4 cores

**Comment by Gaafar** [comment]
@kaicataldo Can you review the testing approach in `tests/lib/cli-engine/cli-engine.js` before I add the remaining cases

**Comment by Gaafar** [comment]
@ilyavolodin @mysticatea is anyone available to review the testing approach in `tests/lib/cli-engine/cli-engine.js`? I'm trying to get your opinion before I manually add another 100+ test cases

**Comment by mysticatea** [comment]
Wow, I apologize that I had not noticed that you are continuing to working on this after the PoC. Please don't advance to implement the unapproved design. PoC is worthful to confirm feasibility, but the advance of work likely goes in a different direction.

Currently, we have been discussing the brand new API for asynchronous on RFC40. We will migrate our APIs to asynchronous entirely and existing `CLIEngine` will be deprecated. Because we have been meeting several big events:

- The `worker_threads` module has arrived stable on Node.js 12.11.0.
- ES modules have arrived stable on 13.2.0
- Async Iteration syntax has arrived stable on 10.0.0, and 8.x will be EOL in November 2019.

Therefore, since we will have the first-class asynchronous APIs and thread supports, (and existing RFCs for parallel are inactive while a long time), so I have written the new design based on the first-class asynchronous APIs and threads: RFC42.

RFC42 is different from RFC4 that it implements parallel linting into the `executeOnFiles()` method directly with the `worker_threads` module and we will enable parallel linting by default in the future.

So we must implement parallel linting based on RFC42. And RFC42 depends on RFC40. And we are in the middle of discussing the design of RFC40 and RFC42. The design is still changeable.

Please participate in the design discussion and wait for the approval.

Thank you!

**Comment by Gaafar** [comment]
Thanks for the update. That's bit of a bummer honestly. I finally have some free time to pick this up and I was hoping to finish the PR by next week.

I had a quick look at RFC42 and it seems very similar to this PR. I can still make the necessary adjustments to this PR without needing to wait for RFC40. Does this make sense?

**Comment by mysticatea** [comment]
RFC42 has many different points from RFC4 and RFC42 has its own implementation that I'm working to aim at 7.0.0 (see the RFC). If you are interested, I hope you to contribute to that implementation. But please mind the design is still changeable.

**Comment by Gaafar** [comment]
I'm traveling for most of this month. If you can ping me once the design is finalized, I'll try to reimplement this accordingly. Thanks

**Comment by nzakas** [comment]
Closing as we have a new design that will make this PR obsolete. Thanks for all of the work, it really pushed the team to make some big changes to enable parallel linting.

**Comment by choheekim** [comment]
@nzakas It'd be nice if you can share a link of rfc or pr that is being worked on to enable parallel liniting?

**Comment by nzakas** [comment]
Please see the previous comments where the RFCs are discussed.

**Pull request closed without merging**